network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/configuration/WgVpnControllerService.kt

```diff
@@ -16,14 +16,87 @@
+import com.duckduckgo.di.scopes.VpnScope
+import com.duckduckgo.mobile.android.vpn.service.VpnSocketProtector
+import com.duckduckgo.networkprotection.impl.di.ProtectedVpnControllerService
+import com.duckduckgo.networkprotection.impl.di.UnprotectedVpnControllerService
+import com.squareup.anvil.annotations.ContributesTo
+import dagger.Lazy
+import dagger.Module
+import dagger.Provides
+import dagger.SingleInstanceIn
+import java.security.KeyStore
+import java.security.Security
+import javax.inject.Named
+import javax.net.ssl.SSLSocket
+import javax.net.ssl.TrustManagerFactory
+import javax.net.ssl.X509TrustManager
+import okhttp3.OkHttpClient
+import org.conscrypt.Conscrypt
+import retrofit2.Retrofit
 import retrofit2.http.Body
 import retrofit2.http.GET
 import retrofit2.http.Headers
 import retrofit2.http.POST
 
+@Module
+@ContributesTo(scope = VpnScope::class)
+object WgVpnControllerServiceModule {
+
+    @Provides
+    @SingleInstanceIn(VpnScope::class)
+    fun provideDelegatingSSLSocketFactory(
+        socketProtector: Lazy<VpnSocketProtector>,
+        @Named("api") okHttpClient: Lazy<OkHttpClient>,
+    ): DelegatingSSLSocketFactory {
+        return object : DelegatingSSLSocketFactory(okHttpClient.get().sslSocketFactory) {
+            override fun configureSocket(sslSocket: SSLSocket): SSLSocket {
+                socketProtector.get().protect(sslSocket)
+                return sslSocket
+            }
+        }
+    }
+
+    @Provides
+    @UnprotectedVpnControllerService
+    @SingleInstanceIn(VpnScope::class)
+    @SuppressLint("NoRetrofitCreateMethodCallDetector")
+    fun providesWgVpnControllerService(
+        @Named(value = "api") retrofit: Retrofit,
+        @Named("api") okHttpClient: Lazy<OkHttpClient>,
+        delegatingSSLSocketFactory: DelegatingSSLSocketFactory,
+    ): WgVpnControllerService {
+        val trustManagerFactory = TrustManagerFactory.getInstance(
+            TrustManagerFactory.getDefaultAlgorithm(),
+        )
+        trustManagerFactory.init(null as KeyStore?)
+        val trustManagers = trustManagerFactory.trustManagers
+        check(!(trustManagers.size != 1 || trustManagers[0] !is X509TrustManager)) {
+            ("Unexpected default trust managers: ${trustManagers.contentToString()}")
+        }
+        val trustManager = trustManagers[0] as X509TrustManager
+
+        val customRetrofit = retrofit.newBuilder()
+            .callFactory {
+                okHttpClient.get().newBuilder()
+                    .sslSocketFactory(delegatingSSLSocketFactory, trustManager)
+                    .build()
+                    .newCall(it)
+            }
+            .build()
+
+        // insertProviderAt trick to avoid error during handshakes
+        Security.insertProviderAt(Conscrypt.newProvider(), 1)
+
+        return customRetrofit.create(WgVpnControllerService::class.java)
+    }
+}
+
 @ContributesServiceApi(AppScope::class)
+@ProtectedVpnControllerService
```

Comment:
Checked this and you are right this doesn’t seem to be necessary.

Reply:
As discussed offline, agreed to remove it